### PR TITLE
Add more e2e tests for DaemonSet templateGeneration and pod adoption

### PR DESCRIPTION
**Pull request by janetkuo**
Depends on #42173

@erictune @kargakis @lukaszo @kubernetes/sig-apps-pr-reviews

### CODE REVIEW

File: test/e2e/daemon_set.go [hunk 1]
```diff
@@ -237,24 +238,32 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
...
 		Expect(err).NotTo(HaveOccurred())
+		Expect(ds.Spec.TemplateGeneration).To(Equal(int64(1)))
 
 		By("Check that daemon pods launch on every node of the cluster.")
 		Expect(err).NotTo(HaveOccurred())
 		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkRunningOnAllNodes(f, label))
 		Expect(err).NotTo(HaveOccurred(), "error waiting for daemon pod to start")
 
+		By("Make sure daemon pods have no template generation set")
+		checkDaemonPodsTemplateGeneration(c, ns, label, "")
```
Comment (lukaszo) [inline test/e2e/daemon_set.go]: TemplateGeneration should always be set. Here it should be set to 1. You don't check if this function returns error.

File: test/e2e/daemon_set.go [hunk 2]
```diff
@@ -237,24 +238,32 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
...
 		Expect(err).NotTo(HaveOccurred())
+		Expect(ds.Spec.TemplateGeneration).To(Equal(int64(2)))
 
-		By("Check that demon pods have not set updated image.")
+		By("Check that demon pods images aren't updated.")
 		err = wait.Poll(dsRetryPeriod, dsRetryTimeout, checkDaemonPodsImage(c, ns, label, image))
 		Expect(err).NotTo(HaveOccurred())
 
+		By("Make sure daemon pods have no template generation set")
+		checkDaemonPodsTemplateGeneration(c, ns, label, "")
```
Comment (lukaszo) [inline test/e2e/daemon_set.go]: ditto

File: test/e2e/daemon_set.go [hunk 3]
```diff
@@ -469,3 +592,64 @@ func checkDaemonPodsImage(c clientset.Interface, ns string, selector map[string]
 		return true, nil
 	}
 }
+
+func templateGenerationMatch(c clientset.Interface, ns string, selector map[string]string, templateGeneration string) func() (bool, error) {
+	return func() (bool, error) {
+		err := checkDaemonPodsTemplateGeneration(c, ns, selector, templateGeneration)
+		match := err == nil
+		return match, nil
```
Comment (lukaszo) [inline test/e2e/daemon_set.go]: You can also return `err` here

**Comment by lukaszo** [comment]
No matter which update strategy is set, label with `TemplateGeneration`  is always added to the daemon pods.

File: test/e2e/daemon_set.go (hunk 1 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: Oh, right. This is fixed

File: test/e2e/daemon_set.go (hunk 2 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: Done

File: test/e2e/daemon_set.go (hunk 3 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: This is the condition func for wait.Poll. If we return an `err` that's not `nil`, we'll immediately exit the wait loop.

**Comment by janetkuo** [comment]
@lukaszo comments address. I also combined the test for setting templateGeneration when creating a DS with the test for RollingUpdate, since they're pretty much the same. PTAL.

File: test/e2e/daemon_set.go (hunk 3 above)
Comment (lukaszo) [inline test/e2e/daemon_set.go]: ah, right

**Comment by lukaszo** [comment]
lgtm :)

**Comment by janetkuo** [comment]
Multiple test builds failed because of connection timed out:

```
W0228 11:21:50.794] + ./hack/install-etcd.sh
W0228 11:23:58.215] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 11:23:58.316] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 4 retries left.
```

@k8s-bot kops aws e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot verify test this
@k8s-bot kubemark e2e test this
@k8s-bot unit test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

**Comment by lukaszo** [comment]
Yeah, amazon s3 went down today.

**Comment by janetkuo** [comment]
@k8s-bot unit test this

**Comment by janetkuo** [comment]
Another unrelated test flake. 

```
k8s.io/kubernetes/test/integration/client TestPatch 2.10s
client_test.go:283: Failed updating patchpod with patch type application/merge-patch+json: Operation cannot be fulfilled on pods "patchpod": there is a meaningful conflict:
	 diff1=[123 34 109 101 116 97 100 97 116 97 34 58 123 34 108 97 98 101 108 115 34 58 110 117 108 108 44 34 114 101 115 111 117 114 99 101 86 101 114 115 105 111 110 34 58 34 49 51 52 51 34 125 125]
	, diff2=[123 34 109 101 116 97 100 97 116 97 34 58 123 34 108 97 98 101 108 115 34 58 123 34 102 111 111 34 58 34 98 97 114 34 125 125 125]
```

@k8s-bot unit test this

File: test/e2e/daemon_set.go [hunk 4]
```diff
@@ -233,28 +234,38 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
...
-		ds, err := c.Extensions().DaemonSets(ns).Get(dsName, metav1.GetOptions{})
+		ds, err = c.Extensions().DaemonSets(ns).Get(dsName, metav1.GetOptions{})
 		ds.Spec.Template.Spec.Containers[0].Image = redisImage
-		_, err = c.Extensions().DaemonSets(ns).Update(ds)
+		ds, err = c.Extensions().DaemonSets(ns).Update(ds)
 		Expect(err).NotTo(HaveOccurred())
+		Expect(ds.Spec.TemplateGeneration).To(Equal(int64(2)))
 
-		By("Check that demon pods have not set updated image.")
+		By("Check that demon pods images aren't updated.")
```
Comment (soltysh) [inline test/e2e/daemon_set.go]: s/demon/daemon

File: test/e2e/daemon_set.go [hunk 5]
```diff
@@ -264,32 +275,118 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
...
+		ds, err = c.Extensions().DaemonSets(ns).Get(dsName, metav1.GetOptions{})
 		ds.Spec.Template.Spec.Containers[0].Image = redisImage
 		ds.Spec.UpdateStrategy = extensions.DaemonSetUpdateStrategy{Type: extensions.RollingUpdateDaemonSetStrategyType}
-		_, err = c.Extensions().DaemonSets(ns).Update(ds)
+		ds, err = c.Extensions().DaemonSets(ns).Update(ds)
 		Expect(err).NotTo(HaveOccurred())
+		Expect(ds.Spec.TemplateGeneration).To(Equal(templateGeneration + 1))
 
-		By("Check that demon pods have not set updated image.")
+		By("Check that demon pods images are updated.")
```
Comment (soltysh) [inline test/e2e/daemon_set.go]: s/demon/daemon

**Review by soltysh — COMMENTED** [review]
Nits, but LGTM.

File: test/e2e/daemon_set.go (hunk 4 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: Fixed

File: test/e2e/daemon_set.go (hunk 5 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: Fixed

**Comment by janetkuo** [comment]
Fixed nits and squashed commits, applying lgtm

**Comment by janetkuo** [comment]
@k8s-bot kops aws e2e test this

**Comment by janetkuo** [comment]
@k8s-bot gce etcd3 e2e test this

File: test/e2e/daemon_set.go [hunk 6]
```diff
@@ -469,3 +566,64 @@ func checkDaemonPodsImage(c clientset.Interface, ns string, selector map[string]
...
+	podList := listDaemonPods(c, ns, label)
+	pods := podList.Items
+	if len(pods) > 1 {
+		sort.Sort(podByCreationTimestamp(pods))
+	}
+	return pods
+}
+
+// podByCreationTimestamp sorts a list of DaemonSet pods by creation timestamp, using their names as a tie breaker.
+type podByCreationTimestamp []v1.Pod
```
Comment (kargakis) [inline test/e2e/daemon_set.go]: Don't we have this already somewhere else?

File: test/e2e/daemon_set.go [hunk 7]
```diff
@@ -264,32 +275,118 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
...
+
+		By(fmt.Sprintf("Make sure all daemon pods have correct template generation %d", templateGeneration))
+		err = checkDaemonPodsTemplateGeneration(c, ns, label, fmt.Sprint(templateGeneration))
+		Expect(err).NotTo(HaveOccurred())
+
+		dsPodsLastCreationTime := getDaemonPodsLastCreationTime(c, ns, label)
+
+		By(fmt.Sprintf("Deleting DaemonSet %s and orphaning its pods", dsName))
+		trueVar := true
+		deleteOptions := &metav1.DeleteOptions{OrphanDependents: &trueVar}
```
Comment (kargakis) [inline test/e2e/daemon_set.go]: Doesn't this depend on owner references?

Comment (janetkuo) [inline test/e2e/daemon_set.go]: yes it depends on #42173

File: test/e2e/daemon_set.go (hunk 6 above)
Comment (janetkuo) [inline test/e2e/daemon_set.go]: We have a `podByCreationTimestamp` in daemonset controller but it's `[]*v1.Pod` not `[]v1.Pod`

**Comment by janetkuo** [comment]
Will wait for #42173 to be merged

**Comment by janetkuo** [comment]
#42173 is now merged. Rebased and re-applying lgtm

**Comment by janetkuo** [comment]
@k8s-bot gci gce e2e test this

**Comment by janetkuo** [comment]
@k8s-bot cvm gce e2e test this

**Comment by janetkuo** [comment]
Another unrelated test failure :/ 
It is too flaky.

@k8s-bot cvm gce e2e test this

**Pull request merged**